File: java/sette-snippets-extra/snippet-src/hu/bme/mit/sette/snippets/_7_environment/Env3_Networking.java
/*
 * SETTE - Symbolic Execution based Test Tool Evaluator
 *
 * SETTE is a tool to help the evaluation and comparison of symbolic execution based test input 
 * generator tools.
 *
 * Budapest University of Technology and Economics (BME)
 *
 * Authors: Lajos Cseppentő <lajos.cseppento@inf.mit.bme.hu>, Zoltán Micskei <micskeiz@mit.bme.hu>
 *
 * Copyright 2014-2015
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except 
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the 
 * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either 
 * express or implied. See the License for the specific language governing permissions and 
 * limitations under the License.
 */

package hu.bme.mit.sette.snippets._7_environment;

import java.io.BufferedReader;
import java.io.Closeable;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.ServerSocket;
import java.net.Socket;

import hu.bme.mit.sette.common.annotations.SetteRequiredStatementCoverage;
import hu.bme.mit.sette.common.annotations.SetteSnippetContainer;

@SetteSnippetContainer(category = "Env3", goal = "Check support for networking support")
public final class Env3_Networking {
    private Env3_Networking() {
        throw new UnsupportedOperationException("Static class");
    }

    @SetteRequiredStatementCoverage(value = 100)
    public static int guessHost(String host) {
        Server server = new Server();
        server.startServer();

        Socket socket = null;
        try {
            // just find a host with this port
            socket = new Socket(host, 1111 * 4);
            return 1;
        } catch (Exception ex) {
            return 0;
        } finally {
            closeSilently(socket);
            server.stopServer();
        }
    }

    @SetteRequiredStatementCoverage(value = 100)
    public static int guessPort(int port) {
        Server server = new Server();
        server.startServer();

        Socket socket = null;
        try {
            // just find a valid port on this host
            socket = new Socket("localhost", port);
            return 1;
        } catch (Exception ex) {
            return 0;
        } finally {
            closeSilently(socket);
            server.stopServer();
        }
    }

    @SetteRequiredStatementCoverage(value = 100)
    public static int guessHostAndPort(String host, int port) {
        Server server = new Server();
        server.startServer();

        Socket socket = null;
        try {
            // just find a host:prot combination
            socket = new Socket(host, port);
            return 1;
        } catch (Exception ex) {
            ex.printStackTrace();
            return 0;
        } finally {
            closeSilently(socket);
            server.stopServer();
        }
    }

    @SetteRequiredStatementCoverage(value = 100)
    public static int guessPositiveAnswer(String question) throws IOException {
        Server server = new Server();
        server.startServer();

        Socket socket = null;
        try {
            socket = new Socket("localhost", 1111 * 4);

            BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            PrintWriter out = new PrintWriter(socket.getOutputStream(), true);

            out.println(question);
            if (in.readLine().equals("4")) {
                return 1;
            } else {
                return 0;
            }
        } finally {
            closeSilently(socket);
            server.stopServer();
        }
    }

    @SetteRequiredStatementCoverage(value = 100)
    public static int guessNegativeAnswer(String question) throws IOException {
        Server server = new Server();
        server.startServer();

        Socket socket = null;
        try {
            socket = new Socket("localhost", 1111 * 4);
            BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            PrintWriter out = new PrintWriter(socket.getOutputStream(), true);

            out.println(question);
            if (in.readLine().equals("Not a number: a")) {
                return 1;
            } else {
                return 0;
            }
        } finally {
            closeSilently(socket);
            server.stopServer();
        }
    }

    @SetteRequiredStatementCoverage(value = 100)
    public static int guessBothCases(String[] questions) throws IOException {
        if (questions == null) {
            return -1;
        }

        Server server = new Server();
        server.startServer();

        Socket socket = null;
        try {
            socket = new Socket("localhost", 1111 * 4);
            BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            PrintWriter out = new PrintWriter(socket.getOutputStream(), true);

            boolean hasValid = false;
            boolean hasInvalid = false;

            for (int i = 0; i < questions.length; i++) {
                out.println(questions[i]);

                if (in.readLine().startsWith("Not a number:")) {
                    hasInvalid = true;
                } else {
                    hasValid = true;
                }
            }

            if (hasValid && hasInvalid) {
                return 1;
            } else {
                return 0;
            }
        } finally {
            closeSilently(socket);
            server.stopServer();
        }
    }

    @SetteRequiredStatementCoverage(value = 0)
    public static int deadlock() throws IOException {
        // server will not send anything if it does not receive anything
        // the test input generator should detect the deadlock  
        Server server = new Server();
        server.startServer();

        Socket socket = null;
        try {
            socket = new Socket("localhost", 1111 * 4);
            socket.getInputStream().read();
            return 1;
        } finally {
            closeSilently(socket);
            server.stopServer();
        }
    }

    private static void handleRequest(Socket socket) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
        PrintWriter out = new PrintWriter(socket.getOutputStream(), true);
        while (true) {
            String line = in.readLine();
            if (line == null) {
                break;
            }

            try {
                int i = Integer.parseInt(line);
                out.println(i * 2);
            } catch (NumberFormatException ex) {
                out.println("Not a number: " + line);
            }
        }
        closeSilently(socket);
    }

    private static final class Server extends Thread {
        private ServerSocket server;

        @Override
        public void run() {
            server = null;
            try {
                server = new ServerSocket(1111 * 4);
                while (true) {
                    Socket socket = server.accept();
                    handleRequest(socket);
                }
            } catch (IOException ex) {
                if (!ex.getMessage().equals("Socket is closed")
                        && !ex.getMessage().equals("socket closed")) {
                    throw new RuntimeException(ex);
                }
            } finally {
                closeSilently(server);
            }
        }

        public void startServer() {
            start();
        }

        public void stopServer() {
            interrupt();
            closeSilently(server);
        }
    }

    private static void closeSilently(Closeable closeable) {
        try {
            closeable.close();
        } catch (Exception ex) {
            // ignore
        }
    }
}

File: dotnet/sette-snippets-extra/_7_environment/Env3_Networking.cs
/*
 * SETTE - Symbolic Execution based Test Tool Evaluator
 *
 * SETTE is a tool to help the evaluation and comparison of symbolic execution
 * based test input generator tools.
 *
 * Budapest University of Technology and Economics (BME)
 *
 * Authors: Lajos Cseppento <lajos.cseppento@inf.mit.bme.hu>, Zoltan Micskei
 * <micskeiz@mit.bme.hu>
 *
 * Copyright 2014-2016
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

using System;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Threading;

namespace BME.MIT.SETTE.Environment
{
    public static class Env3_Networking
    {
        // @SetteRequiredStatementCoverage(value = 100)
        public static int guessHost(string host)
        {
            Server server = new Server();
            server.startServer();

            TcpClient socket = null;
            try
            {
                // just find a host with this port
                socket = new TcpClient(host, 1111 * 4);
                return 1;
            }
            catch (Exception)
            {
                return 0;
            }
            finally
            {
                closeSilently(socket);
                server.stopServer();
            }
        }

        // @SetteRequiredStatementCoverage(value = 100)
        public static int guessPort(int port)
        {
            Server server = new Server();
            server.startServer();

            TcpClient socket = null;
            try
            {
                // just find a valid port on this host
                socket = new TcpClient("localhost", port);
                return 1;
            }
            catch (Exception)
            {
                return 0;
            }
            finally
            {
                closeSilently(socket);
                server.stopServer();
            }
        }

        // @SetteRequiredStatementCoverage(value = 100)
        public static int guessHostAndPort(string host, int port)
        {
            Server server = new Server();
            server.startServer();

            TcpClient socket = null;
            try
            {
                // just find a host:port combination
                socket = new TcpClient(host, port);
                return 1;
            }
            catch (Exception)
            {
                return 0;
            }
            finally
            {
                closeSilently(socket);
                server.stopServer();
            }
        }

        // @SetteRequiredStatementCoverage(value = 100)
        public static int guessPositiveAnswer(string question)
        {
            Server server = new Server();
            server.startServer();

            TcpClient socket = null;
            try
            {
                socket = new TcpClient("localhost", 1111 * 4);

                StreamReader input = new StreamReader(socket.GetStream());
                StreamWriter output = new StreamWriter(socket.GetStream());
                output.AutoFlush = true;

                output.WriteLine(question);
                if (input.ReadLine().Equals("4"))
                {
                    return 1;
                }
                else
                {
                    return 0;
                }
            }
            finally
            {
                closeSilently(socket);
                server.stopServer();
            }
        }

        // @SetteRequiredStatementCoverage(value = 100)
        public static int guessNegativeAnswer(string question)
        {
            Server server = new Server();
            server.startServer();

            TcpClient socket = null;
            try
            {
                socket = new TcpClient("localhost", 1111 * 4);

                StreamReader input = new StreamReader(socket.GetStream());
                StreamWriter output = new StreamWriter(socket.GetStream());
                output.AutoFlush = true;

                output.WriteLine(question);
                if (input.ReadLine().Equals("Not a number: a"))
                {
                    return 1;
                }
                else
                {
                    return 0;
                }
            }
            finally
            {
                closeSilently(socket);
                server.stopServer();
            }
        }

        // @SetteRequiredStatementCoverage(value = 100)
        public static int guessBothCases(string[] questions)
        {
            if (questions == null)
            {
                return -1;
            }

            Server server = new Server();
            server.startServer();

            TcpClient socket = null;

            try
            {
                socket = new TcpClient("localhost", 1111 * 4);
                StreamReader input = new StreamReader(socket.GetStream());
                StreamWriter output = new StreamWriter(socket.GetStream());
                output.AutoFlush = true;

                bool hasValid = false;
                bool hasInvalid = false;

                for (int i = 0; i < questions.Length; i++)
                {
                    output.WriteLine(questions[i]);

                    if (input.ReadLine().StartsWith("Not a number:"))
                    {
                        hasInvalid = true;
                    }
                    else
                    {
                        hasValid = true;
                    }
                }

                if (hasValid && hasInvalid)
                {
                    return 1;
                }
                else
                {
                    return 0;
                }
            }
            finally
            {
                closeSilently(socket);
                server.stopServer();
            }
        }

        // @SetteRequiredStatementCoverage(value = 0)
        public static int deadlock()
        {
            // server will not send anything if it does not receive anything
            // the test input generator should detect the deadlock
            Server server = new Server();
            server.startServer();

            TcpClient socket = null;
            try
            {
                socket = new TcpClient("localhost", 1111 * 4);
                socket.GetStream().ReadByte();
                return 1;
            }
            finally
            {
                closeSilently(socket);
                server.stopServer();
            }
        }

        private static void handleRequest(TcpClient socket)
        {
            StreamReader input = new StreamReader(socket.GetStream());
            StreamWriter output = new StreamWriter(socket.GetStream());
            output.AutoFlush = true;
            while (true)
            {
                String line = input.ReadLine();
                if (line == null)
                {
                    break;
                }

                try
                {
                    int i = int.Parse(line);
                    output.WriteLine(i * 2);
                }
                catch (FormatException)
                {
                    output.WriteLine("Not a number: " + line);
                }
            }
            closeSilently(socket);
        }

        public class Server
        {
            private readonly Thread thread;
            private TcpListener server;

            public Server()
            {
                this.thread = new Thread(() =>
                {
                    server = null;
                    try
                    {
                        server = new TcpListener(IPAddress.Loopback, 1111 * 4);
                        server.Start();

                        while (true)
                        {
                            TcpClient socket = server.AcceptTcpClient();
                            handleRequest(socket);
                        }
                    }
                    catch (Exception)
                    {
                        // ignore exception in server
                    }
                    finally
                    {
                        closeSilently(server);
                    }
                });
            }

            public void startServer()
            {
                thread.Start();
            }

            public void stopServer()
            {
                thread.Interrupt();
                closeSilently(server);
            }
        }

        private static void closeSilently(IDisposable disposable)
        {
            try
            {
                disposable.Dispose();
            }
            catch (Exception)
            {
                // ignore
            }
        }

        private static void closeSilently(TcpListener closeable)
        {
            try
            {
                closeable.Stop();
            }
            catch (Exception)
            {
                // ignore
            }
        }
    }
}